### ISSUE TITLE
whether to reduce loss when tracking step/epoch training loss

### ISSUE THREAD

**Issue by shliu0** [huggingface/accelerate#2109]
I am confused that if i want to track the true loss of each training step/epoch, should I use accelerator.gather_for_metric() like [this](https://github.com/huggingface/accelerate/blob/5b3f3b99d6aa1d71eaa3380af87a6ca6126505fd/examples/complete_cv_example.py#L243) script did for reducing among processes? 
- however, [this](https://github.com/huggingface/accelerate/blob/5b3f3b99d6aa1d71eaa3380af87a6ca6126505fd/examples/complete_cv_example.py#L221C11-L221C11) example script just simply add loss at each step to get epoch loss, can someone help me figure out which method is correct or why different procedures are used to train loss and eval metric seperately? Will averaging loss across all processes make a difference to see whether the training of NN is going well?

- Moreover, if i use multi gpus & gradient_accumulation, the the total_batch_size will be batch_size * num_processes * gradient_accumulation_steps, so the correct formula to get loss of each step should be `loss = accelerator.gather_for_metric(loss).mean().item()`?

- I also see someone use `loss = accelerator.gather(loss.repeat(batch_size)).mean().item()` [link](https://github.com/huggingface/transformers/issues/13247) , is it just an old verison&bug to fix? 

I would be very appreciate if someone can answer me these questions!

**Comment by BenjaminBossan** [comment]
Regarding your first question, I think you're confusing two things. In the first script that you link, the loss is not gathered. What's being gathered are predictions and targets to calculate the evaluation metrics, which are not being used for updating the model, just for reporting metrics. When you have those, you should gather them if you want to have the same metrics on all processes. When it comes to the loss, it may differ on each process, since the training is distributed among the processes.

**Comment by muellerzr** [comment]
True loss will always be the same across processes after `.backward()` iirc. You can check this by printing the `loss` at each step. They should be identical because DDP will average all of them across processes when doing the backward pass

**Comment by ShaowuChen** [comment]
> True loss will always be the same across processes after `.backward()` iirc. You can check this by printing the `loss` at each step. They should be identical because DDP will average all of them across processes when doing the backward pass

Hi muellerzr, I have done some experiments. and it turns out that each process does a batch of training separately, thus the loss is not identical across processes. Is it normal? Could you please take a look at  my code？

I use 2 GPU, bath_size=256, accelerate==0.25.0; the code is 
`def train(net, accelerator, optimizer, scheduler, train_loader):
        for i, (x, y) in enumerate(tqdm(train_loader, disable=not accelerator.is_local_main_process)):
            output = net(x)
            loss1 = torch.nn.CrossEntropyLoss(reduction='mean')(output,y)
            loss2 = torch.nn.CrossEntropyLoss(reduction='sum')(output,y)
            print('before bp '+str(loss1)+' '+str(loss2))
            optimizer.zero_grad()
            accelerator.backward(loss1)
            print('after bw '+str(loss1)+' '+str(loss2))
            optimizer.step()
            scheduler.step()
            print('after op '+str(loss1)+' '+str(loss2))

            print('gather: ' +str(accelerator.gather((loss1,loss2))))
            
            # do something else
            if i==0:
                  break

        return net, accelerator

net, optimizer, train_loader, val_loader, scheduler = accelerator.prepare(
                        net, optimizer, train_loader, val_loader, scheduler)

net, accelerator, scheduler = train(net, accelerator, optimizer, scheduler, train_loader)
`

and the results are
`before bp tensor(0.8804, device='cuda:1', grad_fn=<NllLossBackward0>) tensor(225.3893, device='cuda:1', grad_fn=<NllLossBackward0>)
before bp tensor(0.8187, device='cuda:0', grad_fn=<NllLossBackward0>) tensor(209.5794, device='cuda:0', grad_fn=<NllLossBackward0>)
after bw tensor(0.8187, device='cuda:0', grad_fn=<NllLossBackward0>) tensor(209.5794, device='cuda:0', grad_fn=<NllLossBackward0>)
after bw tensor(0.8804, device='cuda:1', grad_fn=<NllLossBackward0>) tensor(225.3893, device='cuda:1', grad_fn=<NllLossBackward0>)
after op tensor(0.8187, device='cuda:0', grad_fn=<NllLossBackward0>) tensor(209.5794, device='cuda:0', grad_fn=<NllLossBackward0>)
after op tensor(0.8804, device='cuda:1', grad_fn=<NllLossBackward0>) tensor(225.3893, device='cuda:1', grad_fn=<NllLossBackward0>)
gather:(tensor([0.8187, 0.8804], device='cuda:0'), tensor([209.5794, 225.3893], device='cuda:0'))
gather:(tensor([0.8187, 0.8804], device='cuda:1'), tensor([209.5794, 225.3893], device='cuda:1')).
`

**Comment by shliu0** [comment]
> > I am confused that if i want to track the true loss of each training step/epoch, should I use accelerator.gather_for_metric() like [this](https://github.com/huggingface/accelerate/blob/5b3f3b99d6aa1d71eaa3380af87a6ca6126505fd/examples/complete_cv_example.py#L243) script did for reducing among processes?
> > 
> > * however, [this](https://github.com/huggingface/accelerate/blob/5b3f3b99d6aa1d71eaa3380af87a6ca6126505fd/examples/complete_cv_example.py#L221C11-L221C11) example script just simply add loss at each step to get epoch loss, can someone help me figure out which method is correct or why different procedures are used to train loss and eval metric seperately? Will averaging loss across all processes make a difference to see whether the training of NN is going well?
> > * Moreover, if i use multi gpus & gradient_accumulation, the the total_batch_size will be batch_size * num_processes * gradient_accumulation_steps, so the correct formula to get loss of each step should be `loss = accelerator.gather_for_metric(loss).mean().item()`?
> > * I also see someone use `loss = accelerator.gather(loss.repeat(batch_size)).mean().item()` [link](https://github.com/huggingface/transformers/issues/13247) , is it just an old verison&bug to fix?
> > 
> > I would be very appreciate if someone can answer me these questions!
> 
> Hello, I am also very confused about your questions after googling and viewing the limited documents😭. Have you figured out the problems?

I haven't figure out a thorough answer yet. if you want to get an average loss across different processes, then `loss = accelerator.gather_for_metric(loss).mean().item()` should be used. However, I think it's not need to do so if you just want to see if your training is going well correctly, you can just record loss in the main process after each train epoch, and here is a good example you can follow [](https://github.com/huggingface/transformers/blob/df5c5c62ae253055336f5bb0828ca8e3e15ab6bd/examples/pytorch/translation/run_translation_no_trainer.py#L656C1-L656C1) ( hope this may help u, if there is any mistake, welcome to point out :)

**Comment by shliu0** [comment]
> > True loss will always be the same across processes after `.backward()` iirc. You can check this by printing the `loss` at each step. They should be identical because DDP will average all of them across processes when doing the backward pass
> 
> Hi muellerzr, I have done some experiments. and it turns out that each process does a batch of training separately, thus the loss is not identical across processes. Is it normal? Could you please take a look at my code？
> 
> I use 2 GPU, bath_size=256, accelerate==0.25.0; the code is
> 
> ```
> def train(net, accelerator, optimizer, scheduler, train_loader):
>         for i, (x, y) in enumerate(tqdm(train_loader, disable=not accelerator.is_local_main_process)):
>             output = net(x)
>             loss1 = torch.nn.CrossEntropyLoss(reduction='mean')(output,y)
>             loss2 = torch.nn.CrossEntropyLoss(reduction='sum')(output,y)
>             print('before bp '+str(loss1)+' '+str(loss2))
>             optimizer.zero_grad()
>             accelerator.backward(loss1)
>             print('after bw '+str(loss1)+' '+str(loss2))
>             optimizer.step()
>             scheduler.step()
>             print('after op '+str(loss1)+' '+str(loss2))
> 
>             print('gather: ' +str(accelerator.gather((loss1,loss2))))
>             
>             # do something else
>             if i==0:
>                   break
> 
>         return net, accelerator
> 
> net, optimizer, train_loader, val_loader, scheduler = accelerator.prepare(
>                         net, optimizer, train_loader, val_loader, scheduler)
> 
> net, accelerator, scheduler = train(net, accelerator, optimizer, scheduler, train_loader)
> ```
> 
> and the results are
> 
> ```
> before bp tensor(0.8804, device='cuda:1', grad_fn=<NllLossBackward0>) tensor(225.3893, device='cuda:1', grad_fn=<NllLossBackward0>)
> before bp tensor(0.8187, device='cuda:0', grad_fn=<NllLossBackward0>) tensor(209.5794, device='cuda:0', grad_fn=<NllLossBackward0>)
> after bw tensor(0.8187, device='cuda:0', grad_fn=<NllLossBackward0>) tensor(209.5794, device='cuda:0', grad_fn=<NllLossBackward0>)
> after bw tensor(0.8804, device='cuda:1', grad_fn=<NllLossBackward0>) tensor(225.3893, device='cuda:1', grad_fn=<NllLossBackward0>)
> after op tensor(0.8187, device='cuda:0', grad_fn=<NllLossBackward0>) tensor(209.5794, device='cuda:0', grad_fn=<NllLossBackward0>)
> after op tensor(0.8804, device='cuda:1', grad_fn=<NllLossBackward0>) tensor(225.3893, device='cuda:1', grad_fn=<NllLossBackward0>)
> gather:(tensor([0.8187, 0.8804], device='cuda:0'), tensor([209.5794, 225.3893], device='cuda:0'))
> gather:(tensor([0.8187, 0.8804], device='cuda:1'), tensor([209.5794, 225.3893], device='cuda:1'))
> ```

same outputs as yours, from the result, it seems that loss is still different across processes after .backward(), but gradient will be same across processes after .backward()

**Comment by InhyeokYoo** [comment]
> I haven't figure out a thorough answer yet. if you want to get an average loss across different processes, then `loss = accelerator.gather_for_metric(loss).mean().item()` should be used. However, I think it's not need to do so if you just want to see if your training is going well correctly, you can just record loss in the main process after each train epoch, and here is a good example you can follow [link](https://github.com/huggingface/transformers/blob/df5c5c62ae253055336f5bb0828ca8e3e15ab6bd/examples/pytorch/translation/run_translation_no_trainer.py#L656C1-L656C1) ( hope this may help u, if there is any mistake, welcome to point out :)

The loss that recorded by `total_loss += loss.detach().float()` is the loss of main process's training right? Is it alright to get only partial loss instead total loss?
